src/main/java/net/neoforged/neoforge/common/data/DataResourceManager.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.data;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Predicate;
+import net.minecraft.client.resources.ClientPackSource;
+import net.minecraft.client.resources.IndexedAssetSource;
+import net.minecraft.network.chat.Component;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.server.packs.PackLocationInfo;
+import net.minecraft.server.packs.PackType;
+import net.minecraft.server.packs.repository.PackSource;
+import net.minecraft.server.packs.repository.ServerPacksSource;
+import net.minecraft.server.packs.resources.FallbackResourceManager;
+import net.minecraft.server.packs.resources.Resource;
+import net.minecraft.server.packs.resources.ResourceManager;
+import net.neoforged.fml.ModList;
+import net.neoforged.fml.loading.FMLEnvironment;
+import net.neoforged.neoforge.resource.ResourcePackLoader;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Simple {@link ResourceManager} implementation for looking up resources at runtime during data generation.
+ * <p>
+ * This resource manager can be used to lookup any asset/data file at runtime, assuming that the respective mod is loaded.
+ * <p>
+ * Do take note that under server data-gen runs, most of Minecraft's assets will be missing due to {@link ClientPackSource} being client only.
+ *
+ * @see ResourceManager
+ */
+public final class DataResourceManager {
+    private final FallbackResourceManager client = new FallbackResourceManager(PackType.CLIENT_RESOURCES, "minecraft");
+    private final FallbackResourceManager server = new FallbackResourceManager(PackType.SERVER_DATA, "minecraft");
+
+    @ApiStatus.Internal
+    public DataResourceManager(@Nullable String assetIndex, @Nullable File assetsDir) {
+        if (FMLEnvironment.dist.isClient() && assetIndex != null && assetsDir != null)
+            client.push(ClientPackSource.createVanillaPackSource(IndexedAssetSource.createIndexFs(assetsDir.toPath(), assetIndex)));
+
+        server.push(ServerPacksSource.createVanillaPackSource());
+
+        ModList.get().forEachModInOrder(mod -> {
+            var packInfo = new PackLocationInfo("mod/" + mod.getModId(), Component.empty(), PackSource.BUILT_IN, Optional.empty());
+            var modPack = ResourcePackLoader.createPackForMod(mod.getModInfo().getOwningFile()).openPrimary(packInfo);
+
+            client.push(modPack);
+            server.push(modPack);
+        });
```

Comment:
This should filter out the Minecraft "mod" like `ResourcePackLoader.findResourcePacks()` does and likely also needs to filter out the mod for which datagen is being run to avoid including data/assets which will be overwritten by this run (may require adding back the `--existing` parameter).